File: src/main/java/com/indiepost/repository/utils/CriteriaUtils.java
package com.indiepost.repository.utils;

import com.indiepost.dto.post.PostQuery;
import com.indiepost.model.QPost;
import com.querydsl.core.BooleanBuilder;

/**
 * Created by jake on 17. 1. 14.
 */
public interface CriteriaUtils {
    static BooleanBuilder addConjunction(PostQuery query, BooleanBuilder builder) {
        QPost post = QPost.post;

        builder.and(post.status.eq(query.getStatus()));

        if (query.getCategory() != null) {
            builder.and(post.category.slug.eq(query.getCategory().toLowerCase()));
        }
        if (query.getPublishedAfter() != null) {
            builder.and(post.publishedAt.goe(query.getPublishedAfter()));
        }
        if (query.getPublishedBefore() != null) {
            builder.and(post.publishedAt.loe(query.getPublishedBefore()));
        }
        if (query.getModifiedAfter() != null) {
            builder.and(post.modifiedAt.goe(query.getModifiedAfter()));
        }
        if (query.getModifiedBefore() != null) {
            builder.and(post.modifiedAt.loe(query.getModifiedBefore()));
        }
        if (query.getSplash() != null) {
            builder.and(post.isSplash.eq(query.getSplash()));
        }
        if (query.getFeatured() != null) {
            builder.and(post.isFeatured.eq(query.getFeatured()));
        }
        if (query.getPicked() != null) {
            builder.and(post.isPicked.eq(query.getPicked()));
        }
        return builder;
    }
}

File: src/main/kotlin/com/indiepost/repository/utils/CriteriaUtils.kt
package com.indiepost.repository.utils

import com.indiepost.dto.post.PostQuery
import com.indiepost.model.QPost
import com.querydsl.core.BooleanBuilder

/**
 * Created by jake on 17. 1. 14.
 */
object CriteriaUtils {
    fun addConjunction(query: PostQuery, builder: BooleanBuilder): BooleanBuilder {
        val post = QPost.post

        builder.and(post.status.eq(query.status))

        val tagName = query.primaryTag
        if (tagName != null) {
            builder.and(post.primaryTag.name.eq(tagName.toLowerCase()))
        }
        if (query.publishedAfter != null) {
            builder.and(post.publishedAt.goe(query.publishedAfter))
        }
        if (query.publishedBefore != null) {
            builder.and(post.publishedAt.loe(query.publishedBefore))
        }
        if (query.modifiedAfter != null) {
            builder.and(post.modifiedAt.goe(query.modifiedAfter))
        }
        if (query.modifiedBefore != null) {
            builder.and(post.modifiedAt.loe(query.modifiedBefore))
        }
        if (query.splash != null) {
            builder.and(post.isSplash.eq(query.splash!!))
        }
        if (query.featured != null) {
            builder.and(post.isFeatured.eq(query.featured!!))
        }
        if (query.picked != null) {
            builder.and(post.isPicked.eq(query.picked!!))
        }
        return builder
    }

}
